### PR TITLE
Add experimental IsolationForest trainer for ML.NET

### LINKED ISSUE #3043: Add support for Isolation Forests
Isolation forests appear to be one of the best options available for anomaly detection. I was disappointed to see that ML.NET doesn't support this algorithm.

I don't have much experience in this field, but these people do and they seem to agree: http://www.eurecom.fr/en/publication/5334/download/data-publi-5334_2.pdf

One important aspect of isolation forests is that they allow for unsupervised online learning. There are endless applications for such an algorithm. I'm hoping to see this available in ML.NET in the future!

### PR DESCRIPTION
**Pull request by asp2286**
## Related issues
Fixes #3043

# Add Isolation Forest Anomaly Detection Trainer (Experimental)

This PR adds an **Isolation Forest** anomaly detection trainer for ML.NET.  
Isolation Forest (Liu, Ting, Zhou, 2008) is a tree-ensemble algorithm for unsupervised anomaly detection that isolates outliers via random partitioning. It complements existing ML.NET anomaly detectors (e.g., SR-CNN, IID) with a density-agnostic approach.

---

## Motivation
- Provide a widely-used, general-purpose anomaly detection method.  
- Works without strong distribution assumptions.  
- Produces both a continuous anomaly score and a binary label.  
- Achieves parity with popular libraries like scikit-learn.

---

## Design (v1, Experimental)
- **Core engine**: `IsolationForestModel` (pure C#) implements random partitioning trees, scoring, and SHAP-like path contributions.  
- **Pipeline integration**: `IsolationForestTrainer : IEstimator<ITransformer>` appends:
  - `Score` (float, scaled 0–100; higher = more anomalous),
  - `PredictedLabel` (bool), thresholded by `Contamination` or explicit override.  
- **Options**:
  - `Trees`
  - `SampleSize (psi)`
  - `Seed`
  - `Contamination`
  - `ParallelBuild`
  - `ThresholdOverride`

> ⚠️ **Experimental note**: v1 uses `CustomMapping` internally. Models trained with this trainer cannot currently be persisted with `mlContext.Model.Save()`. A follow-up will introduce a proper `IsolationForestTransformer` with save/load and efficient row-mapping.

---

## Usage
```csharp
var pipeline = ml.Transforms.Concatenate("Features", "X1", "X2")
    .Append(new IsolationForestTrainer(new IsolationForestTrainer.Options
    {
        Trees = 200,
        SampleSize = 256,
        Contamination = 0.02
    }));

var model = pipeline.Fit(data);

### CODE REVIEW

**Comment by asp2286** [comment]
@dotnet-policy-service agree


чт, 4 сент. 2025 г. в 22:34, dotnet-policy-service[bot] <
***@***.***>:
